Hard/SubstringwithConcatenationofAllWords.java

```diff
@@ -0,0 +1,80 @@
+import java.util.*;
+
+/**
+ * 30. Substring with Concatenation of All Words   
+ * You are given a string, s, and a list of words, words, that are all of the same length. Find all starting indices of substring(s) in s that is a concatenation of each word in words exactly once and without any intervening characters.
+ * For example, given:
+ * s: "barfoothefoobarman" words: ["foo", "bar"]  You should return the indices: [0,9].
+ * Tags: Hash Table Two Pointers String
+ * Similar Problems: (H) Minimum Window Substring
+ * @author chenshuna
+ */ 
+
+public class SubstringwithConcatenationofAllWords {
+    public static List<Integer> findSubstring(String s, String[] words) {
+        List<Integer> res = new ArrayList<Integer>();
+        if( s==null || words == null || s.length() == 0 || words.length == 0 ) return res;
+        HashMap<String, Integer> wordDict = new HashMap<String, Integer>();
+        int wordLen = words[0].length();
+        for(String word : words){
+            if(wordDict.containsKey(word))
+                wordDict.put(word, wordDict.get(word) + 1);
+            else 
+                wordDict.put(word, 1);
+        }
+        /**
+         * outside loop
+         *  bar/foo/the/foo/bar/man
+         *  b/arf/oot/hef/oob/arm/an
+         *  ba/rfo/oth/efo/oba/rma/n
+         */ 
+        for(int i = 0; i < wordLen; i++){
+            HashMap<String, Integer> sMap = new HashMap<String, Integer>();
```

Comment:
Instead of creating another map and use `count` to compare, you can just create a copy of the wordDict, then reduce count or remove key from the map while you try to find the concatenation.